n_helpers.c

```diff
@@ -84,6 +84,89 @@ NOTE_C_STATIC void setTime(JTIME seconds);
 NOTE_C_STATIC bool timerExpiredSecs(uint32_t *timer, uint32_t periodSecs);
 NOTE_C_STATIC int ytodays(int year);
 
+static const char BINARY_EOP = '\n';
+
+//**************************************************************************/
+/*!
+  @brief  Decode a binary payload received from the Notecard.
+  @param  inBuf The binary payload.
+  @param  inLen The length of the binary payload.
+  @param  outBuf The buffer to write the decoded payload to. This can be the
+                 same address as inBuf, allowing for in-place decoding.
+  @param  outLen On input, holds the length of outBuf. On output, holds the
+                 length of the decoded data.
+  @returns  NULL on success, else an error string pointer.
+*/
+/**************************************************************************/
+const char *NoteBinaryDecode(const uint8_t *inBuf, uint32_t inLen,
+                             uint8_t *outBuf, uint32_t *outLen)
+{
+    if (*outLen < cobsGuaranteedFit(inLen)) {
+        NOTE_C_LOG_ERROR("output buffer too small");
+        return ERRSTR("output buffer too small", c_err);
+    }
+
+    *outLen = cobsDecode((uint8_t *)inBuf, inLen, BINARY_EOP, outBuf);
+
+    return NULL;
+}
+
+//**************************************************************************/
+/*!
+  @brief  Binary encode a buffer to prepare it for transmission to the Notecard.
+  @param  inBuf The data to encode.
+  @param  inLen The length of the data to encode.
+  @param  outBuf The buffer to write the encoded data to. This can be the
+                 same address as inBuf, allowing for in-place encoding.
+  @param  outLen On input, holds the length of outBuf. On output, holds the
+                 length of the encoded data.
+  @returns  NULL on success, else an error string pointer.
+*/
+/**************************************************************************/
+const char *NoteBinaryEncode(const uint8_t *inBuf, uint32_t inLen,
+                             uint8_t *outBuf, uint32_t *outLen)
+{
+    if (*outLen < cobsEncodedMaxLength(inLen)) {
```

Comment:
It's dangerous to dereference a pointer (especially on a public API) without first checking it for `NULL`.

Reply:
I agree. I just don't see us coding defensively like that in note-c, historically-speaking, but I think it would be good to start. I'll add checks.